### ISSUE TITLE
Missed type narrowing in `assert len(r) == …` which should imply `r` is not None

### ISSUE THREAD

**Issue by jmarshall** [microsoft/pyright#11365]
In the following code, after the first `assert` it would seem that the type of `r` could be narrowed to `list[int]`, which would prevent the pyright diagnostic on the second `assert`:

```python
def foo() -> list[int] | None:
    return [28, 37]

def bar():
    r = foo()
    reveal_type(r)  # list[int] | None
    assert len(r) == 2
    reveal_type(r)  # list[int] | None   but could be   list[int]
    assert 28 in r  # error: Operator "in" not supported for types … and "list[int] | None"
```

**Describe the solution you’d like**

Use of `len(r)` within the assert implies that afterwards `r` cannot be None, so its type can be narrowed to `list[int]`. With that information, the types for the `in` operator are appropriate and so there is no type error.

(This could be worked around with `assert r and len(r) == 2` but it would be good if the existing code were diagnostic-free too.)

**Comment by rchiodo** [comment]
Thanks for the issue but Pyright wouldn't implement something like this. It would require way too much understanding of what len(r) == 2 means. 

Pyright would work with this assert though.

Code sample in [pyright playground](https://pyright-play.net/?code=CYUwZgBGD20BQEoIFoB8EA2BLAzgFwG0sA7PAXQgB8IA5aYkALgCgI2IAnEPAVw%2BIgEATAA4ANBADMAdjLNmoSACMAhh0Qt2nCAF4osRK3ZcAbiBUYA%2BngCeABxBwOSCAGJMuQiXJVa9EEZsKjg4IBx42rgQxNARdAyBnCBmFtb2js5s7tj4RKQU1PEg7Eo8EQDG0DwYwBBKxWw5XvmJwaHhEKIQJNpuEGEc0ByMEADyDhwqeEMQAEQks9GxEDg8dnZDeCC1MBwQtg44EIBkBBAqxLWzTXk%2Bhf6zzEA)

```python
def foo() -> list[int] | None:
    return [28, 37]

def bar():
    r = foo()
    reveal_type(r)  # list[int] | None
    assert r is not None
    reveal_type(r)  # list[int] 
    assert 28 in r  # No more error

```

**Comment by jmarshall** [comment]
Fair enough.

Nonetheless I'm surprised: I would have thought that “in an `assert`, `len(r)` invoked (or many other built-in functions) implies `r` is not None” would be exactly the sort of rule pyright could implement.

At present, I am dealing with a code base in which 90% of the type errors are of the form “type X expected, but got X | None [and don't realise that in reality it can't be None]”. I would expect this is very common, and excellent None-narrowing would be a hallmark of an excellent type checker.
